### PR TITLE
Iceberg: support row-level delete and update

### PR DESCRIPTION
**Pull request by jackye1995**
This PR adds support for writing Iceberg position delete. Similar to #8534 , I first present our working internal implementation backported to Trino, some parts might not work because of internal differences, but once we agree with the general approach I will make the fix and add unit tests. 

Also, there is a missing piece that has to be added after #8534 is first merged, so that the `IcebergPageSource` has the ability to retain the row position channel and pass it to the updatable page source.

A few key points:
1. we choose to support row level delete through position delete spec instead of equality delete, this is based on the general guidance from Iceberg that position delete is preferred when possible. Plus the delete write mechanism in Trino fits perfectly with the position delete spec, as data is first scanned and filtered anyway, recording equality wastes the computation effort comparing to recording positions.
2. Trino delete row ID column has Trino type `ROW(string file_path, long pos, row(table schema))`, which matches Iceberg's position delete file schema.
3. update and delete share exactly the same row id type and also `beginXXX` and `finishXXX` operation implementation. The only difference is that update writes new data files after writing the delete files. This is because update in Iceberg is modelled as delete + insert.
4. In the page source provider, if it detects the operation is for DELETE or UPDATE (columns contain row id column), it will automatically read all the table columns (except for the identity partition columns), because the entire row is a part of the position delete schema anyway.
5. I directly reused the current Iceberg sink implementation to write back the position delete rows and updated rows. It's probably not the most optimal, but simple enough for the first iteration.

This is a bare minimum backport, I left some inline TODOs, and also there are many optimizaitons we can make after the base version is checked in, I tried to keep this as simple as possible to avoid too many disagreements around optimization related changes. Please let me know if this looks good or not, thanks!

@phd3 @electrum @findepi @losipiuk @caneGuy @rdblue @hashhar

### CODE REVIEW

**Comment by findepi** [comment]
@jackye1995 can you please add product test that would assert compatibility between Trino and Spark?
see https://github.com/trinodb/trino/blob/1364fbbcd0b001bfe0d51ebd88244aa661796b12/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java#L1911 for how we test Trino/Hive compatibility for ORC ACID tables. This approach has proven to be useful and helped find some bugs too.

**Comment by lhofhansl** [comment]
If I wanted to try to this out, I'd need to create an Iceberg Table adhering to the Iceberg Format Specification V2, since you are proposing using delete snapshots, right?

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlock.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ private RowBlock(int startOffset, int positionCount, @Nullable boolean[] rowIsNu
     }
 
     @Override
-    protected Block[] getRawFieldBlocks()
+    public Block[] getRawFieldBlocks()
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlock.java:115]: Wonder why this is needed, and whether this is actually used  correctly.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java [hunk 2]
```diff
@@ -116,4 +122,9 @@ public static IcebergColumnHandle create(Types.NestedField column, TypeManager t
                 toTrinoType(column.type(), typeManager),
                 Optional.ofNullable(column.doc()));
     }
+
+    public static IcebergColumnHandle createUpdateRowIdColumnHandle(Schema tableSchema, TypeManager typeManager)
+    {
+        return create(required(ROW_ID_COLUMN_INDEX, ROW_ID_COLUMN_NAME, DeleteSchemaUtil.posDeleteSchema(tableSchema).asStruct()), typeManager);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:128]: Is it used for deletes only, or for updates as well?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 3]
```diff
@@ -255,6 +258,11 @@ public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
                 name.getTableName(),
                 name.getTableType(),
                 snapshotId,
+                getFileFormat(table),
+                getDataPath(table.location()),
+                serializeToBytes(table.schema()),
+                serializeToBytes(table.spec()),
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:264]: I think there already was an idea to add schema to IcebergTableHandle and it was rejected (?) for some reason.

@phd3 do you remember?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 4]
```diff
@@ -176,7 +205,45 @@ public ConnectorPageSource createPageSource(
...
 
-        return new IcebergPageSource(icebergColumns, partitionKeys, dataPageSource, session.getTimeZoneKey());
+        ConnectorPageSource icebergPageSource = new IcebergPageSource(icebergColumns, partitionKeys, dataPageSource, session.getTimeZoneKey());
+
+        if (!hasDeleteRowId) {
+            return icebergPageSource;
+        }
+        else {
+            Schema posDeleteSchema = DeleteSchemaUtil.posDeleteSchema(table.getSchema());
+            ConnectorPageSink posDeleteSink = new IcebergPageSink(
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:215]: positionalDeletesSink

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+public class IcebergPageSourceUpdate
+        implements UpdatablePageSource
+{
+    private final Slice filePathSlice;
+    private final List<IcebergColumnHandle> queriedColumns;
+    private final List<IcebergColumnHandle> allTableColumns;
+    private final List<IcebergColumnHandle> updateColumns;
+    private final ConnectorPageSource source;
+    private final ConnectorPageSink posDeleteSink;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java:47]: positionalDeletesSink

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 6]
```diff
@@ -176,7 +205,45 @@ public ConnectorPageSource createPageSource(
...
+                    hdfsEnvironment,
+                    hdfsContext,
+                    icebergColumns,
+                    jsonCodec,
+                    session,
+                    table.getFileFormat(),
+                    FileContent.POSITION_DELETES,
+                    maxOpenPartitions);
+
+            ConnectorPageSink updateRowSink = new IcebergPageSink(
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:230]: updatedDataSink

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    private final ConnectorPageSink updateRowSink;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java:48]: updatedDataSink

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        deleteRows(rowIdBlock);
+        Map<Integer, Block> updatedColumnBlocks = new HashMap<>();
+        for (int i = 0; i < columnValueAndRowIdChannels.size() - 1; i++) {
+            IcebergColumnHandle updatedColumn = updateColumns.get(i);
+            Block updatedValues = page.getBlock(columnValueAndRowIdChannels.get(i));
+            updatedColumnBlocks.put(updatedColumn.getId(), updatedValues);
+        }
+
+        Block[] updatedRows = new Block[allTableColumns.size()];
+        Block[] oldRows = ((RowBlock) rowIdBlock.getRawFieldBlocks()[2]).getRawFieldBlocks();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java:89]: Cast to `RowBlock` isn't entirely correct.
See #9354 and perhaps we should use `ColumnarRow` here. 
cc @djsstarburst

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        rowIdComponentBlocks[2] = RowBlock.fromFieldBlocks(pageSize, Optional.empty(), fieldBlocks);
+
+        Block[] resultBlocks = new Block[queriedColumns.size()];
+        for (int i = 0; i < queriedColumns.size(); i++) {
+            IcebergColumnHandle columnHandle = queriedColumns.get(i);
+            if (ROW_ID_COLUMN_NAME.equals(columnHandle.getName())) {
+                resultBlocks[i] = RowBlock.fromFieldBlocks(pageSize, Optional.empty(), rowIdComponentBlocks);
+            }
+            else {
+                resultBlocks[i] = sourcePage.getBlock(allTableColumns.indexOf(columnHandle));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java:142]: `indexOf` use here looks quadratic, and we seem to be doing this for every page.

**Review by findepi — COMMENTED** [review]
cc @alexjo2144

**Comment by jackye1995** [comment]
close in favor of #10075

**Pull request closed without merging**